### PR TITLE
Add SymbolicShape and replace all uses of VaryingShape<ShapeSymbol> with it

### LINKED ISSUE #38411: Switch `c10::optional<ShapeSymbol>` with just `ShapeSymbol`.
## 🚀 Feature
* Simplify the `symbolic_sizes_` property of TensorType API. `TensorType` is a class to represent a tensor type in TorchScript IR. `TensorType` tracks various properties of a tensor such as tensor shape, strides, device, requires_grad, etc.  TorchScript IR consists mostly of `Node` and `Value` . `Node`s denote operations that produce outputs (`Value`s) . Each value is statically typed with `Type` and that type can be `TensorType` .
*  Currently, an unknown size is expressed with [VaryingShape](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/core/jit_type.h#L423) that wraps around a vector of `c10::optional<int64_t>`s
* We recently added `ShapeSymbol` class to represent `size` and values and it has its own way of expressing unknown values`
    * ShapeSymbol now provides a special symbol (`-1`) which we can use to denote unknown symbols (https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/core/jit_type.h#L389) 
 * In short, we would like to switch from using optionals to denote unknown symbols to using that special symbol.
## Motivation

* The outcomes we hope for
    * The reduction in the amount of boxing/unboxing
    * Avoiding having more than one way to express the same thing.

## Pitch
* We would like to add `unknownSymbol` to denote unknown/varying shape
* We would like to use `unknownSymbol` in lieu of `c10::optional<ShapeSymbol>`
* We would like to replace `VaryingShape<ShapeSymbol>` with just `c10::optional<std::vector<ShapeSymbol>>` 

## Alternatives

Do nothing n and suffer from the current design.

## Additional context

For questions, ping @Krovatkin 

cc @suo

### PR DESCRIPTION
**Pull request by gmagogsfm**
Adding a SymbolicShape class to represent a generic tensor shape with ShapeSymbols.

Its core data structure is c10::optional<std::vector<ShapeSymbol>>. If has_value() == false, it represents an unranked tensor shape. At any dimension ShapeSymbol can contain dynamic size, checkable with ShapeSymbol::IsStatic method.

SymbolicShape now replaces all uses of VaryingShape<ShapeSymbol>, ie c10::optional<std::vector<c10::optional<ShapeSymbol>>>. The inner c10::optional wrapper around ShapeSymbol used to indicate dynamic shape, which overlaps with part of ShapeSymbol's representation.

### CODE REVIEW

**Comment by Krovatkin** [comment]
@gmagogsfm This looks good! Do you think we might be able to get rid of `SymbolicShape` as well or is there still too much logic that needs some place to live? We can do it in another PR? Our goal is to reduce the number of helper classes that don't do too much.

**Review by Krovatkin — COMMENTED** [review]
@gmagogsfm This looks good! Do you think we might be able to get rid of SymbolicShape as well or is there still too much logic that needs some place to live? We can do it in another PR? Our goal is to reduce the number of helper classes that don't do too much.

**Comment by gmagogsfm** [comment]
Thanks for taking a look @Krovatkin . I think keeping SymbolicShape still has some value. Based on my limited understanding, it can capture the relationship between different SymbolicShapes, (say same batch dim but different size for other dims). What do you think?

**Comment by gmagogsfm** [comment]
@Krovatkin I noticed that you added another mergeSymbolicShapes function in profiling_record with slightly different logic. Do you think it makes sense the consolidate it together with the one in SymbolicShape class?

**Review by Krovatkin — APPROVED** [review]
:shipit:

**Comment by Krovatkin** [comment]
@gmagogsfm `mergeSymbolicShapes` doesn't really do much in the latest `master` and `partitionSetByDimension` seems to be significantly different. But if you see an opportunity for reuse, go for it!
